### PR TITLE
cpp: added to_lowercase and to_uppercase to SharedString

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Review by ogoffart — COMMENTED** [review]
You can't pass a pointer to an already initialized SharedString in C++ and write it with ptr::write.

Either 
 1. you do not initialize the string before (by making sure its constructor is not called, or by being in the constructor)
 2. you use, in rust, a `&mut SharedString` and you just assign to it
 3. you call the drop function manually before re-initializing it.
 
I'd recommend the option 2 in this case.

File: api/cpp/tests/datastructures.cpp [hunk 1]
```diff
@@ -52,6 +52,17 @@ SCENARIO("SharedString API")
...
+
+    SECTION("to_lowercase")
+    {
+        str = "Hello";
+        REQUIRE(std::string_view(str.to_lowercase().data()) == "hello");
+    }
+    SECTION("to_uppercase")
+    {
+        str = "Hello";
+        REQUIRE(std::string_view(str.to_uppercase().data()) == "hello");
```
Comment (tronical) [inline api/cpp/tests/datastructures.cpp:64]: That's going to be a condition difficult to fulfil ;-)

File: api/cpp/include/slint_string.h [hunk 2]
```diff
@@ -127,6 +127,34 @@ struct SharedString
...
     static SharedString from_number(double n) { return SharedString(n); }
 
+    /// Returns the lowercase equivalent of this string, as a new SharedString.
+    ///
+    /// For example:
+    /// \code
+    ///     auto str = slint::SharedString("Hello");
+    ///     auto str2 = str.to_lowercase(); // creates "hello"
+    /// \endcode
+    SharedString to_lowercase()
```
Comment (tronical) [inline api/cpp/include/slint_string.h:137]: This and the other function should be `const`.

File: api/cpp/include/slint_string.h [hunk 3]
```diff
@@ -127,6 +127,34 @@ struct SharedString
...
+    }
+
+    /// Returns the uppercase equivalent of this string, as a new SharedString.
+    ///
+    /// For example:
+    /// \code
+    ///     auto str = slint::SharedString("Hello");
+    ///     auto str2 = str.to_uppercase(); // creates "HELLO"
+    /// \endcode
+    const SharedString to_uppercase()
```
Comment (tronical) [inline api/cpp/include/slint_string.h:151]: The function needs to be const, not the returned object :)

```suggestion
    SharedString to_lowercase() const
    {
        auto out = SharedString();
        cbindgen_private::slint_shared_string_to_lowercase(&out, this);
        return out;
    }

    /// Returns the uppercase equivalent of this string, as a new SharedString.
    ///
    /// For example:
    /// \code
    ///     auto str = slint::SharedString("Hello");
    ///     auto str2 = str.to_uppercase(); // creates "HELLO"
    /// \endcode
    SharedString to_uppercase() const
```

**Review by tronical — APPROVED** [review]

**Pull request merged**